features/cellular/framework/targets/UBLOX/N2XX/UBLOX_N2XX_CellularStack.cpp

```diff
@@ -0,0 +1,249 @@
+    sock_id = _at.read_int();
+    _at.resp_stop();
+
+    if ((_at.get_last_error() != NSAPI_ERROR_OK) || (sock_id == -1)) {
+        _at.unlock();
+        return NSAPI_ERROR_NO_SOCKET;
+    }
+    _at.unlock();
+
+    // Check for duplicate socket id delivered by modem
+    for (int i = 0; i < N2XX_MAX_SOCKET; i++) {
+        CellularSocket *sock = _socket[i];
+        if (sock && sock->created && sock->id == sock_id) {
+            return NSAPI_ERROR_NO_SOCKET;
+        }
+    }
+
+    socket->id = sock_id;
+    socket->created = true;
+
+    return NSAPI_ERROR_OK;
+}
+
+nsapi_size_or_error_t UBLOX_N2XX_CellularStack::socket_sendto_impl(CellularSocket *socket, const SocketAddress &address,
+        const void *data, nsapi_size_t size)
+{
+    if (size > N2XX_MAX_PACKET_SIZE) {
+        return NSAPI_ERROR_PARAMETER;
+    }
+
+    int sent_len = 0, len = 0;
+    char *dataStr = (char *) malloc((size * 2) + 1);
+    memset(dataStr, 0, size*2+1);
+    char_str_to_hex_str((const char*)data, size, dataStr);
+    len = strlen(dataStr);
+
+    _at.cmd_start("AT+NSOST=");
+    _at.write_int(socket->id);
+    _at.write_string(address.get_ip_address());
+    _at.write_int(address.get_port());
+    _at.write_int(len/2);
+    _at.write_string(dataStr);
+    _at.cmd_stop();
+
+    _at.resp_start();
+    _at.skip_param(); // skip socket id
+    sent_len = _at.read_int();
+    _at.resp_stop();
+
+    free(dataStr);
+    if ((_at.get_last_error() == NSAPI_ERROR_OK)) {
+        return sent_len;
+    }
+
+    return _at.get_last_error();
+}
+
+nsapi_size_or_error_t UBLOX_N2XX_CellularStack::socket_recvfrom_impl(CellularSocket *socket, SocketAddress *address,
+        void *buffer, nsapi_size_t size)
+{
+    nsapi_size_or_error_t nsapi_error_size = NSAPI_ERROR_DEVICE_ERROR;
+    nsapi_size_t read_blk, usorf_sz, count = 0, length = size*2;
+    bool success = true;
+    char ipAddress[NSAPI_IP_SIZE];
+    char *tmpBuf = (char *) malloc(size*2+1);
+    memset(tmpBuf, 0, size*2+1);
+    int port = 0;
+    Timer timer;
+
+    if (socket->pending_bytes == 0) {
+        _at.process_oob();
+        if (socket->pending_bytes == 0) {
+            free(tmpBuf);
+            return NSAPI_ERROR_WOULD_BLOCK;
+        }
+    }
+
+    timer.start();
+    while (success && (length > 0)) {
+        read_blk = N2XX_MAX_PACKET_SIZE;
+        if (read_blk > length) {
+            read_blk = length;
+        }
+        if (socket->pending_bytes > 0) {
+            _at.cmd_start("AT+NSORF=");
+            _at.write_int(socket->id);
+            _at.write_int(read_blk);
+            _at.cmd_stop();
+
+            _at.resp_start();
+            _at.skip_param(); // receiving socket id
+            _at.read_string(ipAddress, sizeof(ipAddress));
+            port = _at.read_int();
+            usorf_sz = _at.read_int();
+
+            // Must use what +NSORF returns here as it may be less or more than we asked for
+            if (usorf_sz >= socket->pending_bytes) {
```

Comment:
If the modem has received more than it has indicated by `NSONMI_URC` then this is true and stops reading, so it's probably more safe to clear `pending_bytes` after reading up to `length`? In that case actual value of `pending_bytes` is not meaningful but it can be used just as a flag to indicate that some amount of data is waiting to be read.